File: Android/shared/src/main/kotlin/shared/dto/enums/ProfessionalEmployee.kt
@file:UseSerializers(URISerializer::class)

package shared.dto.enums

import kotlinx.serialization.DeserializationStrategy
import kotlinx.serialization.Serializable
import kotlinx.serialization.UseSerializers
import kotlinx.serialization.json.JsonContentPolymorphicSerializer
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.jsonObject
import shared.dto.objects.Professional
import shared.dto.protocols.Responsable
import shared.serializers.URISerializer
import java.net.URI

@Serializable(ProfessionalEmployee.Serializer::class)
sealed class ProfessionalEmployee : Responsable {
    @Serializable
    data class Link(
        val link: URI,
    ) : ProfessionalEmployee()

    @Serializable
    data class Value(
        val value: Professional.Responses.Partial,
    ) : ProfessionalEmployee()

    internal object Serializer :
        JsonContentPolymorphicSerializer<ProfessionalEmployee>(ProfessionalEmployee::class) {
        override fun selectDeserializer(element: JsonElement): DeserializationStrategy<ProfessionalEmployee> {
            return when (element.jsonObject.contains("link")) {
                true -> Link.serializer()
                false -> Value.serializer()
            }
        }
    }
}

File: Node/src/dto/enums/ProfessionalEmployee.ts
import { Professional } from "../objects/Professional";
import { ValidateNested } from "class-validator";

export type ProfessionalEmployee = ProfessionalEmployee.Link   | 
                                   ProfessionalEmployee.Value

export namespace ProfessionalEmployee {

    export class Link {
        link: URL

        constructor(link: URL) {
            this.link = link
        }
    }

    export class Value {
        @ValidateNested()
        value: Professional.Responses.Partial

        constructor(value: Professional.Responses.Partial) {
            this.value = value
        }
    }
}